sdks/java/harness/src/main/java/org/apache/beam/fn/harness/data/PCollectionConsumerRegistry.java

```diff
@@ -305,4 +338,59 @@ public double getProgress() {
       return delegate.getProgress();
     }
   }
+
+  private static class SampleByteSizeEstimateDistribution<T> {
+    /** Basic implementation of {@link ElementByteSizeObserver} for use in size estimation. */
+    private static class ByteSizeObserver extends ElementByteSizeObserver {
+      private long observedSize = 0;
+
+      @Override
+      protected void reportElementSize(long elementSize) {
+        observedSize += elementSize;
+      }
+    }
+
+    final Distribution distribution;
+
+    public SampleByteSizeEstimateDistribution(Distribution distribution) {
+      this.distribution = distribution;
+    }
+
+    public void tryUpdate(T value, Coder<T> coder) throws Exception {
+      if (shouldSampleElement() || coder.isRegisterByteSizeObserverCheap(value)) {
+        // First try using byte size observer
+        ByteSizeObserver observer = new ByteSizeObserver();
+        coder.registerByteSizeObserver(value, observer);
+
+        if (!observer.getIsLazy()) {
+          observer.advance();
+          this.distribution.update(observer.observedSize);
+        } else {
+          // Coder byte size observation is lazy (requires iteration for observation) so fall back
+          // to counting output stream
+          CountingOutputStream os = new CountingOutputStream(ByteStreams.nullOutputStream());
+          coder.encode(value, os);
+          this.distribution.update(os.getCount());
+        }
+      }
+    }
+
+    // Lowest sampling probability: 0.001%.
+    private static final int SAMPLING_TOKEN_UPPER_BOUND = 1000000;
+    private static final int SAMPLING_CUTOFF = 10;
+    private int samplingToken = 0;
+    private Random randomGenerator = new Random();
+
+    private boolean shouldSampleElement() {
```

Comment:
Copied sampling logic from legacy worker -- do we want to be able to customize or update these values?

Reply:
It's ok to me to keep the same logic here. Would be really helpful if we have benchmark here.

Reply:
Ran this micro benchmark and got: 
    With coder 79309931
    Without coder 76130704

```
PCollectionConsumerRegistry consumers =
        new PCollectionConsumerRegistry(
            metricsContainerRegistry, mock(ExecutionStateTracker.class));
    FnDataReceiver<WindowedValue<String>> consumerA1 = mock(FnDataReceiver.class);
    consumers.register(pCollectionA, pTransformIdA, consumerA1, StringUtf8Coder.of());
    FnDataReceiver<WindowedValue<String>> wrapperConsumer =
        (FnDataReceiver<WindowedValue<String>>)
            (FnDataReceiver) consumers.getMultiplexingConsumer(pCollectionA);
    String elementValue = "test pcollection size estimation benchmark";
    WindowedValue<String> element = valueInGlobalWindow(elementValue);
    long coderDur = 0;
    int numElements = 1000;
    int runs = 100;
    for (int run = 0; run < runs; run++) {
      long startTime = System.nanoTime();
      for (int i = 0; i < numElements; i++) {
        wrapperConsumer.accept(element);
      }
      long endTime = System.nanoTime();
      coderDur += (endTime - startTime);
    }
    long nullCoderDur = 0;
    consumers.register(pCollectionB, pTransformIdB, consumerA1, null);
    wrapperConsumer =
        (FnDataReceiver<WindowedValue<String>>)
            (FnDataReceiver) consumers.getMultiplexingConsumer(pCollectionB);
    for (long run = 0; run < runs; run++) {
      long startTime = System.nanoTime();
      for (int i = 0; i < numElements; i++) {
        wrapperConsumer.accept(element);
      }
      long endTime = System.nanoTime();
      nullCoderDur += (endTime - startTime);
    }
    System.out.println("With coder " + coderDur/runs);
    System.out.println("Without coder " + nullCoderDur/runs);